python/packages/core/agent_framework/azure/_search_provider.py

```diff
@@ -0,0 +1,538 @@
+                endpoint=endpoint,
+                credential=credential,
+            )
+
+        self._knowledge_base_initialized = False
+
+    async def __aenter__(self) -> Self:
+        """Async context manager entry."""
+        return self
+
+    async def __aexit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_val: BaseException | None,
+        exc_tb: Any,
+    ) -> None:
+        """Async context manager exit - cleanup handled by client destructors.
+
+        Args:
+            exc_type: Exception type if an error occurred.
+            exc_val: Exception value if an error occurred.
+            exc_tb: Exception traceback if an error occurred.
+
+        Note:
+            We don't explicitly close the Azure SDK clients here because doing so
+            can cause "transport already closed" errors on subsequent uses within
+            the same context. The clients will clean up their resources when they
+            are garbage collected.
+        """
+        pass
+
+    @override
+    async def invoking(
+        self,
+        messages: ChatMessage | MutableSequence[ChatMessage],
+        **kwargs: Any,
+    ) -> Context:
+        """Retrieve relevant context from Azure AI Search before model invocation.
+
+        Args:
+            messages: User messages to use for context retrieval.
+            **kwargs: Additional arguments (unused).
+
+        Returns:
+            Context object with retrieved documents as messages.
+        """
+        # Extract query from messages
+        messages_list = [messages] if isinstance(messages, ChatMessage) else list(messages)
+        query = "\n".join(msg.text for msg in messages_list if msg and msg.text and msg.text.strip())
+
+        if not query:
+            return Context()
+
+        # Perform search based on mode
+        if self.mode == "semantic":
+            search_results = await self._semantic_search(query)
+        else:  # agentic
+            search_results = await self._agentic_search(query)
+
+        # Format results as context
+        if not search_results:
+            return Context()
+
+        context_text = f"{self.context_prompt}\n\n{search_results}"
+
+        return Context(messages=[ChatMessage(role="system", text=context_text)])
+
+    async def _semantic_search(self, query: str) -> str:
+        """Perform semantic hybrid search with semantic ranking.
+
+        This is the recommended mode for most use cases. It combines:
+        - Vector search (if embedding_function provided)
+        - Keyword search (BM25)
+        - Semantic reranking (if semantic_configuration_name provided)
+
+        Args:
+            query: Search query text.
+
+        Returns:
+            Formatted search results as string.
+        """
+        vector_queries = []
+
+        # Generate vector query if embedding function provided
+        if self.embedding_function and self.vector_field_name:
```

Comment:
In many cases (including most indexes created from the portal wizard, etc.) vector fields will have vectorizers setup already. You could try and discovery those automatically and not require an embedding endpoint at all for those indexes. Sort of how it's done in Azure MCP: https://github.com/microsoft/mcp/blob/33360d19e352f0e4108088c751faeb185c069897/tools/Azure.Mcp.Tools.Search/src/Services/SearchService.cs#L112